### PR TITLE
 #2244 StreamDefinitions should be immutable since it is a json object

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2244)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefUtils.java [hunk 1]
```diff
@@ -265,76 +249,74 @@
...
+              StreamDefUtils.NODE_NAMES_DEFINITION,
+              StreamDefUtils.VERSION_DEFINITION,
+              StreamDefUtils.REVISION_DEFINITION,
+              StreamDefUtils.AUTHOR_DEFINITION,
+              StreamDefUtils.TAGS_DEFINITION,
+              StreamDefUtils.JAR_KEY_DEFINITION,
+              StreamDefUtils.JAR_INFO_DEFINITION)
+          .collect(Collectors.toList());
+
+  private List<SettingDef> configDefs = new ArrayList<>();
```
Comment (chia7712) [inline ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefUtils.java]: The name "utils", normally, is a collection of (static) helper methods/members. Having the inner non-static variable is weird.

There are two behaviors for this class. first is to show the default definitions. seconds is to parse the value according to definitions. My two cents is that we can have two classes for them.

Comment (saivirtue) [reply ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefUtils.java]: OK

**Comment by saivirtue** [comment]
retry

File: docs/src/user_guide.rst [hunk 2]
```diff
@@ -695,7 +695,7 @@ running any Ohara containers, you need to install Kubernets first.
 We’ll walk you through this process with a few k8s commands:
 
 Install distribute mode for Kubernetes
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (chia7712) [inline docs/src/user_guide.rst]: the purpose of this change?

File: ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java [hunk 3]
```diff
@@ -327,14 +67,15 @@ public StreamDefinitions add(SettingDef config) {
    */
   public StreamDefinitions addAll(List<SettingDef> configList) {
```
Comment (chia7712) [inline ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java]: If the setters are not necessary, making the StreamDefintitions immutable is always a good idea.

File: docs/src/user_guide.rst (hunk 2 above)
Comment (saivirtue) [reply docs/src/user_guide.rst]: The rst plugin of Intellij suggest that. It seems less warning in IDE but want hurt the doc compile. :)

File: ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java (hunk 3 above)
Comment (saivirtue) [reply ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java]: Thanks for the suggestion, I will try to use immutable object instead.

File: docs/src/user_guide.rst (hunk 2 above)
Comment (chia7712) [reply docs/src/user_guide.rst]: Got it

**Comment by saivirtue** [comment]
retry connector

File: ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java [hunk 4]
```diff
@@ -16,338 +16,80 @@
...
-          .key("streamMetricGroup")
-          .group(CORE_GROUP)
-          .orderInGroup(ORDER_COUNTER.getAndIncrement())
-          .optional("streamapp")
-          .readonly()
-          .internal()
-          .valueType(Type.STRING)
-          .build();
+  StreamDefinitions(List<SettingDef> configDefs) {
+    this.configDefs = configDefs;
```
Comment (chia7712) [inline ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java]: a nit: copy the input list and wrap it to a immutable list.

Comment (saivirtue) [reply ohara-streams/src/main/java/com/island/ohara/streams/config/StreamDefinitions.java]: :)

**Pull request merged**